fiftyone/core/odm/dataset.py

```diff
@@ -480,21 +493,109 @@ class DatasetDocument(Document):
     brain_methods = DictField(ReferenceField(RunDocument))
     evaluations = DictField(ReferenceField(RunDocument))
 
+    def get_saved_views(self):
+        saved_views = self.saved_views
+
+        #
+        # When mongoengine encounters a ReferenceField that contains an
+        # ObjectId with no matching document, the doc is a falsey DBRef.
+        #
+        # We're not currently sure why this happens... but we've observed that
+        # ObjectIDs in DatasetDocument can somehow get out of sync with their
+        # corresponding view documents. This attempts to patch on-the-fly.
+        #
+        if any(isinstance(doc, DBRef) for doc in saved_views):
+            try:
+                patch_saved_views(self.name)
```

Comment:
wrapping a potential db write in the read method might lead to teams bugs

Reply:
> I really don't think we should be deleting these automatically. It hides the issue and will make it even harder to debug


It *fixes* the issue, not *hides*. These utils bring the dataset doc back into sync with the corresponding documents in the views/runs collections. If there are no docs in the external collections, there's no point in having an orphan `ObjectId` in the dataset doc.

Reply:
Hmm, its a valid point that this introduces the possibility of an edit operation being attempted when a user is only trying to do a read operation...

Reply:
I can update to instead log a warning telling the user they should manually run the relevant `patch_XXX()` utils.

Reply:
^updated